### PR TITLE
[MRG] add `sig fileinfo` and standard manifest-generating functionality

### LINKED ISSUE #1190: sourmash index describe (or similar)
It would be nice to have a function(maybe `sourmash index describe`) that reports index characteristics (scaled, ksize, moltype, sbt version, storage type?, etc).

We can currently get this info by running `sourmash sig describe | head`, and inferring from the signature info.

>@luizirber (on slack): `sourmash index describe` also opens the possibility of making `sourmash index` into a larger subcommand (like `sourmash sig`), but then I think we need to move current functionality into `sourmash index build` or something like that

### LINKED ISSUE #1439: provide a summarization command for signature collections
from https://github.com/dib-lab/sourmash/pull/1349#issuecomment-786850141 - 

@bluegenes suggests:

>`sourmash sig describe` --> yes, would be nice to support usage, but output may not be as useful? Likely too many sigs for full output, and `sourmash sig describe <zipfile> | head` would no longer represent the parameters of all sigfiles, since it's possible to incorporate incompatible signatures. If enabled, maybe we could summarize the sig params (e.g. num sigs per ksize/alpha/scaled ) as tsv or tabbed terminal output? Or have a quick default option that gives info for first sig, and `summarize` option that gives full summary? Somewhat related to #1190.

maybe a `sourmash sig summary` or `sourmash sig summarize` command?

### PR DESCRIPTION
**Pull request by ctb**
Inspired by the discussion in https://github.com/sourmash-bio/sourmash/pull/1830, this PR adds `sourmash sig fileinfo` for summary information on a path, and uses that functionality as inspiration to add generic `get_manifest(...)` internal functionality.

Specifically, this PR:
* adds `sourmash sig fileinfo` which provides a useful human-readable summary of a path; see below for example output.
* provides a new standard function `get_manifest(idx, *, require=True, rebuild=False)` that can be used to get a manifest for any index object.
* refactors `extract` to use manifests, which should be (much) faster for databases
* updates the `Index` base class to have `__len__`

In addition, this PR:
* adds `__len__()` and manifest-generating support to `LCA_Database`
* documents `sourmash_args` a bit in the module docstring
* fixes `MultiIndex.location` to work properly in all circumstances
* adds manifest support to `ZipFileLinearIndex.__len__`, which speeds it up tremendously
* fixes a general problem with loading manifest column `with_abundance` from a CSV
* signatures from stdin are now put in a `MultiIndex` class with manifest support, instead of just a `LinearIndex`; this means that all sourmash CLI functions now use Index classes with manifests
* adds `__eq__` to manifest objects so that `m1 == m2` is based on content equality, not object equality

Fixes #1190
Fixes #1439

## Example output

```
% == This is sourmash version 4.2.5.dev19+g3a6028fb.d20220217. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

** loading from 'tests/test-data/prot/all.zip'
path filetype: ZipFileLinearIndex
location: /Users/t/dev/sourmash/tests/test-data/prot/all.zip
is database? yes
has manifest? yes
num signatures: 8
** examining manifest...
31758 total hashes
summary of sketches:
   2 sketches with dayhoff, k=19, scaled=100          7945 total hashes
   2 sketches with hp, k=19, scaled=100               5184 total hashes
   2 sketches with protein, k=19, scaled=100          8214 total hashes
   2 sketches with DNA, k=31, scaled=1000             10415 total hashes
```

## Limitations

`sourmash sig extract` will not work when used with `--picklist` on certain database types (LCA DBs, SBTs, and zipfiles w/o manifests). This is because we now convert manifests into picklists to do the extraction and these database types do not support multiple picklists. See https://github.com/sourmash-bio/sourmash/issues/1848 for one possible resolution.

IMO this is a pretty minor limitation so I decided not to complicate this PR by fixing it. Let me know if you disagree :).

See test `test_sig_extract_8_picklist_md5_lca` for a test that demonstrates this behavior.

## Questions

* should we use `sourmash sig summarize` instead of, or in addition to, `fileinfo`? I'm not sure `fileinfo` is the first place I'd look for this functionality, but it _is_ very descriptive 😆 
* do we want a human-readable "filetype" that's more informative for e.g. .sig files and directories? Right now it just says "MultiIndex" which doesn't really tell you what was loaded (since MultiIndex is used for many things)

## TODO

- [x] provide standard argparse functions to force rebuild of manifest
- [x] figure out what idx.location is so often None!? (for .sig files and directories)
- [x] also check is empty/Index.bool and Index.len stuff; update Index spec?
- [ ] add YAML output to `fileinfo`
- [x] add tests for `get_manifest(...)` on various class types
- [x] add tests for `fileinfo` on all the index types
- [x] add tests for `MultiIndex.location` fix
- [x] add tests for `LCA_Database` use of `__len__` and new manifest capability
- [x] add tests for `ZipFileLinearIndex` proper use of manifests + select
- [x] add documentation for different database types?
- [x] mine the comments for issues
- [x] expand the docstring at top of sourmash_args
- [x] provide information on combinations of sketches
- [x] test `extract` on database types that don't support multiple rounds of picklists
- [x] look at stdin loading

### CODE REVIEW

**Comment by ctb** [comment]
note from chatting with @bluegenes : the combinatorial information would be good (this ksize,scaled,moltype => this many sketches)

**Comment by ctb** [comment]
why is location empty for `sourmash sig fileinfo podar-ref/1.fa.sig`

**Comment by ctb** [comment]
maybe the thing to do is to speed up extract as part of this, with reference to #1830

**Comment by ctb** [comment]
do we want to add a 'get' method or something to a Index classes to retrieve specific signatures by internal location?

**Comment by ctb** [comment]
```
% cat tests/test-data/47.fa.sig | sourmash sig fileinfo -

== This is sourmash version 4.2.5.dev19+g3a6028fb.d20220217. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

** loading from '-'
path filetype: LinearIndex
location: <_io.TextIOWrapper name='<stdin>' mode='r' encoding='utf-8'>
is database? no
has manifest? no
is nonempty? yes
num signatures: 1
** examining manifest...
Generating a manifest...
ERROR: manifests cannot be generated for <_io.TextIOWrapper name='<stdin>' mode='r' encoding='utf-8'>
```

**Comment by ctb** [comment]
ok @bluegenes I've got the stdout on this looking pretty much like I want it, only a few more tests and output options to go before I put it up for merge. if you have time to look at the stdout to make suggestions, then please do so! no hurry, tho.

This PR turned out to an interesting exercise in implementation that found some minor-but-frustrating bugs in how we were passing around file paths, and using picklists and manifests. Very useful!

**Comment by ctb** [comment]
Note to all: this adds yaml as a sourmash requirement. I like yaml output, but @luizirber has made the point that YAML is ill-specified elsewhere. Maybe we should make this optional - if you have it, you can use `sourmash sig fileinfo --yaml-out ...`, if you don't, you can't.

@sourmash-bio/devs ready for initial review! But not merge yet.

**Comment by ctb** [comment]
removed pyyaml altogether, as an unnecessary complication. Ready for review!

File: src/sourmash/sourmash_args.py
Comment (bluegenes) [inline src/sourmash/sourmash_args.py]:
```suggestion
    If a manifest exists and `rebuild` is False, return the manifest.
```

File: tests/test_cmd_signature_fileinfo.py
Comment (bluegenes) [inline tests/test_cmd_signature_fileinfo.py]:
```suggestion
def test_sig_fileinfo_does_not_exist(runtmp):
```

File: doc/command-line.md
Comment (bluegenes) [inline doc/command-line.md]: i think you now also provide total # hashes for each sketch type?

File: doc/command-line.md
Comment (bluegenes) [inline doc/command-line.md:1013]: all other info is `label: #`, rather than `# label`.

Doesn't really matter - looks fine this way too 🤷🏻‍♀️

File: doc/command-line.md
Comment (bluegenes) [inline doc/command-line.md:1012]: do manifests have an associated creation date/datetime that we could print? Perhaps this is more related to manifest metadata (#1847)?

**Review by bluegenes — APPROVED** [review]
lgtm. Left a few minor comments/suggestions.

Quite excited about fileinfo, more unified manifest functionality, and manifest-based db sig extraction 🎉

**Comment by bluegenes** [comment]
Closes #1190, #1439

File: doc/command-line.md
Comment (ctb) [reply doc/command-line.md:1012]: they do not, alas.

File: doc/command-line.md
Comment (ctb) [reply doc/command-line.md:1013]: > Closes #1190, #1439

hah, hadn't even noticed 🤣

**Comment by ctb** [comment]
>Closes #1190, #1439

hah! Hadn't even noticed!

File: doc/command-line.md
Comment (ctb) [reply doc/command-line.md:1013]: fixed in 56a678e7, thx! consistency is good, tho people should use JSON instead of parsing this output 👀

File: doc/command-line.md
Comment (ctb) [reply doc/command-line.md]: yep! good catch. fixed in d4facce0

**Pull request merged**
